File: concrete/src/main/java/pcb/uwu/amount/quantity/Minutes.java
package pcb.uwu.amount.quantity;

import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.quantity.TimeUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.quantity.MinuteUnit.MINUTE;

public class Minutes extends Time {

	// region constructors

	public Minutes(Number value) {
		super(value, MINUTE);
	}

	public Minutes(Number value, Magnitude magnitude) {
		super(value, magnitude, MINUTE);
	}

	public Minutes(String value) {
		super(value, MINUTE);
	}

	public Minutes(String value, Magnitude magnitude) {
		super(value, magnitude, MINUTE);
	}

	public Minutes(BigDecimal value) {
		super(value, MINUTE);
	}

	public Minutes(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, MINUTE);
	}

	public Minutes(BigDecimalAmount amount) {
		super(amount, MINUTE);
	}

	public Minutes(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, MINUTE);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Minutes plus(UnitAmount<TimeUnit> other, MathContext mathContext) {
		return new Minutes(plusAmount(other, mathContext));
	}

	@Override
	public Minutes minus(UnitAmount<TimeUnit> other, MathContext mathContext) {
		return new Minutes(minusAmount(other, mathContext));
	}

	@Override
	public Minutes multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Minutes(multipliedByScalar(other, mathContext));
	}

	@Override
	public Minutes dividedBy(BigDecimal other, MathContext mathContext) {
		return new Minutes(dividedByScalar(other, mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/base/Minutes.kt
package pcb.uwu.amount.base

import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.MinuteUnit
import pcb.uwu.unit.base.TimeUnit

class Minutes : Time
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, MinuteUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, MinuteUnit)

    // region UnitAmount

    override fun plus(time: UnitAmount<TimeUnit>) =
        Minutes(this.amount + (time to MinuteUnit).amount)

    override fun minus(time: UnitAmount<TimeUnit>) =
        Minutes(this.amount + (time to MinuteUnit).amount)

    override fun times(scalar: Number) =
        Minutes(this.amount * scalar)

    override fun div(scalar: Number) =
        Minutes(this.amount / scalar)

    // endregion
}
